merge_command_test.go

```diff
@@ -197,158 +200,184 @@ var _ = TestWebhookHandler(func(context WebhookTestContext) {
-								mockMergeFailWithConflict()
-							})
-
-							It("retries and fails with a gateway error", func() {
-								handle()
-
-								// +1 because of the initial attempt
-								pullRequests.AssertNumberOfCalls(GinkgoT(), "Get", grh.MergeRetryLimit+1)
-								pullRequests.AssertNumberOfCalls(GinkgoT(), "Merge", grh.MergeRetryLimit+1)
-								Expect(responseRecorder.Code).To(Equal(http.StatusBadGateway))
-							})
-						})
-
-						Context("with merge succeeding with first retry", func() {
-							BeforeEach(func() {
-								mockMergeFailWithConflict().Once()
-
-								additionalCommitMessage := ""
-								pullRequests.
-									On(
-										"Merge",
-										repositoryOwner,
-										repositoryName,
-										issueNumber,
-										additionalCommitMessage,
-										noSquashOpts,
-									).
-									Return(&github.PullRequestMergeResult{
-										Merged: github.Bool(true),
-									}, nil, nil).
-									Once()
-							})
-
-							It("removes the 'merging' label from the PR after the merge", func() {
-								issues.
-									On("RemoveLabelForIssue", repositoryOwner, repositoryName, issueNumber, grh.MergingLabel).
-									Return(nil, nil, nil)
-
-								handle()
-								pullRequests.AssertNumberOfCalls(GinkgoT(), "Get", 2)
-								pullRequests.AssertNumberOfCalls(GinkgoT(), "Merge", 2)
-								Expect(responseRecorder.Code).To(Equal(http.StatusOK))
-							})
-						})
-					})
+var ItMergesPR = func(context WebhookTestContext, issueAuthor string, issueNumber int) {
+	var (
+		handle = context.Handle
+
+		responseRecorder *httptest.ResponseRecorder
+		pullRequests     *mocks.PullRequests
+		issues           *mocks.Issues
+	)
+	BeforeEach(func() {
+		responseRecorder = *context.ResponseRecorder
+		pullRequests = *context.PullRequests
+		issues = *context.Issues
+	})
 
-					Context("with merge failing, because PR not mergeable", func() {
-						BeforeEach(func() {
-							additionalCommitMessage := ""
-							resp := &http.Response{
-								StatusCode: http.StatusMethodNotAllowed,
-							}
-							pullRequests.
-								On(
-									"Merge",
-									repositoryOwner,
-									repositoryName,
-									issueNumber,
-									additionalCommitMessage,
-									noSquashOpts,
-								).
-								Return(nil, &github.Response{
-									Response: resp,
-								}, &github.ErrorResponse{
-									Response: resp,
-									Message:  "Pull Request is not mergeable",
-								}).
-								Once()
-						})
-
-						It("fails with a gateway error", func() {
-							handle()
-							Expect(responseRecorder.Code).To(Equal(http.StatusBadGateway))
-						})
-					})
+	Context("with merge failing with an unknown error", func() {
+		BeforeEach(func() {
+			additionalCommitMessage := ""
+			pullRequests.
+				On(
+					"Merge",
+					repositoryOwner,
+					repositoryName,
+					issueNumber,
+					additionalCommitMessage,
+					noSquashOpts,
+				).
+				Return(nil, nil, errors.New("an error")).
```

Comment:
>This would add too much noise to the tests

I believe this what I'm proposing:
```go
Return(NO_RESULT, NO_RESPONSE, errors.New("an error")).
```

That's not noise - that's information provided to the reader vs `nil`. With just `nil`, the reader must remember the method signature. Be kind to the reader - he has a lot to keep in his head already.

>Creating names for all of the return values would add, on average, 2 lines to every mock setup.

Can't they be constants scoped beyond each test? Unless I'm missing something, you only need to define them once for the entire suite.